_pytest/cacheprovider.py

```diff
@@ -149,7 +149,9 @@ def pytest_sessionfinish(self, session):
         config = self.config
         if config.getvalue("cacheshow") or hasattr(config, "slaveinput"):
             return
-        config.cache.set("cache/lastfailed", self.lastfailed)
+        prev_failed = config.cache.get("cache/lastfailed", None) is not None
+        if (session.items and prev_failed) or self.lastfailed:
```

Comment:
Session.items only exists on xdist slaves and normal test runs

A master process does not run collection